# 2023-03-09

Cilium의 Hubble 설치 시 gRPC 통신이 안된다고 안됨
다시 설치 진행

- kubernetes 1.22 환경 워커노드 3대
- cilium 1.10 버전

관련 URL
- https://docs.cilium.io/en/v1.10/gettingstarted/hubble_setup/


(1) cilium 으로 설치하기 -> helm 으로 설치한게 아니어서 secret 에러남 ㅠㅠ
- 관련 이슈 : https://github.com/cilium/cilium-cli/issues/959
```
$ cilium hubble enable
⚠️  Error parsing helm cli secret: unable to retrieve helm values secret kube-system/cilium-cli-helm-values: secrets "cilium-cli-helm-values" not found
⚠️  Proceeding in unknown installation state
🔮 Auto-detected cilium version v1.10.16
🔑 Created CA in secret cilium-ca
ℹ️  helm template --namespace kube-system cilium cilium/cilium --version 1.10.16 --set hubble.enabled=true,hubble.relay.enabled=true,hubble.tls.ca.cert=LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUNGRENDQWJxZ0F3SUJBZ0lVQlhoYWlsd3VIWkNCOTYvSDNzbk5EMS9IWHZJd0NnWUlLb1pJemowRUF3SXcKYURFTE1Ba0dBMVVFQmhNQ1ZWTXhGakFVQmdOVkJBZ1REVk5oYmlCR2NtRnVZMmx6WTI4eEN6QUpCZ05WQkFjVApBa05CTVE4d0RRWURWUVFLRXdaRGFXeHBkVzB4RHpBTkJnTlZCQXNUQmtOcGJHbDFiVEVTTUJBR0ExVUVBeE1KClEybHNhWFZ0SUVOQk1CNFhEVEl6TURNd09UQXlNalF3TUZvWERUSTRNRE13TnpBeU1qUXdNRm93YURFTE1Ba0cKQTFVRUJoTUNWVk14RmpBVUJnTlZCQWdURFZOaGJpQkdjbUZ1WTJselkyOHhDekFKQmdOVkJBY1RBa05CTVE4dwpEUVlEVlFRS0V3WkRhV3hwZFcweER6QU5CZ05WQkFzVEJrTnBiR2wxYlRFU01CQUdBMVVFQXhNSlEybHNhWFZ0CklFTkJNRmt3RXdZSEtvWkl6ajBDQVFZSUtvWkl6ajBEQVFjRFFnQUVNVk5XbGtXaTNQMGRQR2NXbElLejF3SEcKOXB2TjN4Z0hEZXFuRFhDWnNCdGJodzJmalZHZk5mY1EvRVk4bzdoMStOQmJXTGc1a21xQ0N6MGoyN3dNS3FOQwpNRUF3RGdZRFZSMFBBUUgvQkFRREFnRUdNQThHQTFVZEV3RUIvd1FGTUFNQkFmOHdIUVlEVlIwT0JCWUVGQlZ2Cmo5OE9nNlU2MkRHNFBkS0FzcUhDWWlOVE1Bb0dDQ3FHU000OUJBTUNBMGdBTUVVQ0lRQ3czc1dJQXNFMSt4RlcKTThHeUp3SnpJeFRHRmpTSDhSL3kxa2VITmd1eVd3SWdEY0ZmenRaZU1kRGRTOTB6MGp3ajFyVXBZRzNGNThiOQpUM0VFdVZXMVIyTT0KLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo=,hubble.tls.ca.key=[--- REDACTED WHEN PRINTING TO TERMINAL (USE --redact-helm-certificate-keys=false TO PRINT) ---]
✨ Patching ConfigMap cilium-config to enable Hubble...
🚀 Creating ConfigMap for Cilium version 1.10.16...
♻️  Restarted Cilium pods
⌛ Waiting for Cilium to become ready before deploying other Hubble component(s)...
```

(번외) 일단 hubble 설치
```
$ export HUBBLE_VERSION=$(curl -s https://raw.githubusercontent.com/cilium/hubble/master/stable.txt)
$ curl -L --remote-name-all https://github.com/cilium/hubble/releases/download/$HUBBLE_VERSION/hubble-linux-amd64.tar.gz{,.sha256sum}
$ sha256sum --check hubble-linux-amd64.tar.gz.sha256sum
$ sudo tar xzvfC hubble-linux-amd64.tar.gz /usr/local/bin
$ rm hubble-linux-amd64.tar.gz{,.sha256sum}
```

(2) apply 로 설치 (성공-> 한 클러스터에서만..)
```
$ kubectl apply -f https://raw.githubusercontent.com/cilium/cilium/v1.9/install/kubernetes/quick-install.yaml
$ kubectl apply -f https://raw.githubusercontent.com/cilium/cilium/v1.9/install/kubernetes/quick-hubble-install.yaml
```

로그 보면
```
level=info msg=Connecting address="192.168.129.8:4244" hubble-tls=true peer=nks-7403-m-2eew subsys=hubble-relay
level=warning msg="Failed to create gRPC client" address="192.168.129.6:4244" error="connection error: desc = \"transport: error while dialing: dial tcp 192.168.129.6:4244: connect: connection refused\"" hubble-tls=true next-try-in=1h30m0s peer=nks-7403-m-2eeu subsys=hubble-relay
level=info msg=Connecting address="192.168.129.6:4244" hubble-tls=true peer=nks-7403-m-2eeu subsys=hubble-relay
level=warning msg="Failed to create gRPC client" address="192.168.129.7:4244" error="connection error: desc = \"transport: error while dialing: dial tcp 192.168.129.7:4244: connect: connection refused\"" hubble-tls=true next-try-in=1h30m0s peer=nks-7403-m-2eev subsys=hubble-relay
level=info msg=Connecting address="192.168.129.7:4244" hubble-tls=true peer=nks-7403-m-2eev subsys=hubble-relay
level=warning msg="Failed to create gRPC client" address="192.168.129.8:4244" error="connection error: desc = \"transport: error while dialing: dial tcp 192.168.129.8:4244: connect: connection refused\"" hubble-tls=true next-try-in=1h30m0s peer=nks-7403-m-2eew subsys=hubble-relay
level=warning msg="Failed to create gRPC client" address="192.168.129.6:4244" error="connection error: desc = \"transport: error while dialing: dial tcp 192.168.129.6:4244: connect: connection refused\"" hubble-tls=true next-try-in=1h30m0s peer=nks-7403-m-2eeu subsys=hubble-relay
level=warning msg="Failed to create gRPC client" address="192.168.129.7:4244" error="connection error: desc = \"transport: error while dialing: dial tcp 192.168.129.7:4244: connect: connection refused\"" hubble-tls=true next-try-in=1h30m0s peer=nks-7403-m-2eev subsys=hubble-relay
level=info msg=Connecting address="192.168.129.10:4244" hubble-tls=true peer=node-1-w-2h65 subsys=hubble-relay
level=info msg=Connecting address="192.168.129.11:4244" hubble-tls=true peer=node-1-w-2h66 subsys=hubble-relay
level=info msg=Connecting address="192.168.129.9:4244" hubble-tls=true peer=node-1-w-2eex subsys=hubble-relay
level=info msg=Connected address="192.168.129.10:4244" hubble-tls=true peer=node-1-w-2h65 subsys=hubble-relay
level=info msg=Connected address="192.168.129.11:4244" hubble-tls=true peer=node-1-w-2h66 subsys=hubble-relay
level=info msg=Connected address="192.168.129.9:4244" hubble-tls=true peer=node-1-w-2eex subsys=hubble-relay
```

릴레이가 성공하는 애가 있고 아닌 애가 있는데 보면 마스터임 ㅡㅡㅋ

(번외) hubble ui 확인
```
while true; do cilium connectivity test; done
```


왜 안됐냐?
hubble-ui의 backend 로그 중 이런 애가 있음
```
level=error msg="fetching hubble flows: connecting to hubble-relay (attempt #17) failed: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial tcp: lookup hubble-relay on 169.254.25.10:53: no such host\"\n" subsys=ui-backend
```

gma..
coredns에서 넣었던 설정 빼니까 됨 -> hosts 이게 문젠데

```
apiVersion: v1
data:
  Corefile: |
    .:53 {
        errors
        health {
            lameduck 5s
        }
        ready
        hosts {
          ip archive.com
        }
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        forward . /etc/resolv.conf {
          prefer_udp
        }
        cache 30

        loop
        reload
        loadbalance
    }
kind: ConfigMap
metadata:
  labels:
    addonmanager.kubernetes.io/mode: EnsureExists
  name: coredns
  namespace: kube-system
```

쪼끔 더 우아하게 로그 보기
```
$ k logs -f -n kube-system --selector 'k8s-app=kube-dns'
```

---


- kubernetes crd
Custom Resource Definition (CRD)

kubernetes에서 지원하는 리소스 이외의 새로운 리소스를 정의하고 관리할 수 있음

---

cilium cluster mesh 설치
-> https://docs.cilium.io/en/v1.9/gettingstarted/clustermesh/
```
Prerequisites
- PodCIDR ranges in all clusters must be non-conflicting.
```
전제 조건에 걸림
파드 대역은 겹치면 안됨 ㅠ.ㅠ 그래서 안됩니다~